### PR TITLE
feat: properly handle task limits

### LINKED ISSUE #28: Properly handle task limit
The original implementation allowed for productive contributors to work on two tasks concurrently. With a starting two task limit:
| Condition                                              | Task Adjustment       | Explanation                                                                                             |
|--------------------------------------------------------|-----------------------|---------------------------------------------------------------------------------------------------------|
| If any reviewer approved the pull                      | +1 task per pull      | It's likely that the pull is good, just waiting on other slow reviewers to confirm.                      |
| If reviewers take longer than 24 hours                 | +1 task per pull      | Don't wait for lazy reviewers to be able to start a new task.                                            |
| If any reviewer requested changes                      | -1 task per pull      | They should focus on addressing the changes as a top priority.                                           |

_Originally posted by @0x4007 in https://github.com/ubiquibot/command-start-stop/pull/19#discussion_r1731441502_
            

For the sake of this conversation a completed PR is when the conditions are met in the above table.

In the current state the code checks if any review is approved or if there is 0 reviews but 24 hours have passed since the creation of the PR, however it doesn't check if there's any requested changes so it will count it as completed even if there 1 approve and 1 requested changes.

There is still a problem when the reviewer requests changes and the contributor resolves those changes but waits for the reviewer to make a new review so they can't start another task.
There's two possible solutions: check if the reviewer was requested more than x hours ago or check that changes have been marked as solved

### PR DESCRIPTION
**Pull request by jordan-ae**
Resolves #28

This PR updates the conditions for handling task limits

### CODE REVIEW

**Comment by 0x4007** [comment]
@zugdev perhaps you can help test and review this since you seem the most affected

**Comment by zugdev** [comment]
> @zugdev perhaps you can help test and review this since you seem the most affected

sure!

**Comment by zugdev** [comment]
jordan do you have a bot running this fix?

**Comment by jordan-ae** [comment]
> sure! @jordan-ae do you have a bot running this fix?

I'll add a QA link to the issue preview soon. But to test this locally you'll have to set the [ubiquity-os-kernel](https://github.com/ubiquity-os/ubiquity-os-kernel) up, clone and setup the [command-start-stop repo](https://github.com/ubiquity-os-marketplace/command-start-stop). Then you can test this locally

**Comment by jordan-ae** [comment]
@0x4007 I've added a QA. Looks like everyone is really busy but the changes here are really minimal should be pretty straight forward to review hopefully

**Comment by 0x4007** [comment]
@Whilefoo @gentlementlegen @zugdev can you test this

**Comment by zugdev** [comment]
With your QA you should've demonstrated that if a PR is waiting for review the active issue count of the user is aliviated. That's so he can start more issues while his PRs are pending review. I'll try to QA it myself tomorrow or the day after that.

**Comment by gentlementlegen** [comment]
What should also be tested is the +1 task available after the threshold for reviews is due.

**Comment by jordan-ae** [comment]
@zugdev & @gentlementlegen  Note sure if I linked it poorly but that's what the QA demonstrates. The user's initial task limit is 2. He self assigns the two task and can't asign himself to task 3 until an approved pr is opened against one of his current task. Once a pr is opened and has an approval +1 task is added to his task limit.

**Comment by gentlementlegen** [comment]
I tested with your changes and here is what I get:
- I am using user @ubiquity-ubiquibot
- It has one Issue assigned: https://github.com/Meniole/command-start-stop/issues/39
- This issue has a PR assigned: https://github.com/Meniole/command-start-stop/pull/42
- The task limit for the contributors is set to `1`
- the review delay is set to `1 minute`
- it tries to start another task: https://github.com/Meniole/command-start-stop/issues/43

Finally the assigment gets denied with the following values
```
        ⚠ You have reached your max task limit
        ⚠ {
            "assignedIssues": 1,
            "openedPullRequests": 0,
            "limit": 1,
            "caller": "handleTaskLimitChecks"
          }
```

It seems the `openedPullRequest` is wrong.

**Comment by jordan-ae** [comment]
Screenshot from 2024-11-03 17-11-02

@gentlementlegen I've been facing this error on /start when testing. It's trying to check pull request that do not exist I've cleared all existing changes and still get the same error. Do you have any idea what might be causing the error?

**Comment by gentlementlegen** [comment]
@jordan-ae never happened to me before. Have you changed your code for testing and forgot to revert it? It should be fetching the pull-requests from GitHub API so there cannot be items that do not exist there.

**Comment by jordan-ae** [comment]
Thought so too. But I've reverted all the changes, same Issue. So I recloned the repo, same issue still.

**Comment by gentlementlegen** [comment]
Does this reference a private repository? Otherwise there could be a bug then you should investigate why this repository gets listed if it doesn't exist.

File: src/utils/issue.ts
Comment (jordan-ae) [inline src/utils/issue.ts:238]: @gentlementlegen I think the reason why I keep getting a 404 is that `openedPullRequests` returns the openIssues for the particular user and not pull request information. When mapping through through `openedPullRequests`. We use `openedPullRequest.number` as the pull request number whereas it's the issue number that's why it returns a `404` for that pull because it doesn't exist. 

When I was testing coincidentally I had an issue opened with the same number as the pull request I used for testing reason why I didnt get the error I guess. 

What's on this branch for `openedPullRequests` is same as whats on development. So I'm not sure how this ever worked to begin with.

Comment (gentlementlegen) [reply src/utils/issue.ts:238]: Then we can add the `is:pr` filter to this line
https://github.com/ubiquity-os-marketplace/command-start-stop/blob/9152118c14d0a9efb4a2b32ee969b62117aab73e/src/utils/issue.ts#L179
Let me know if that resolves your problem.

Comment (gentlementlegen) [reply src/utils/issue.ts:238]: @jordan-ae I just understand now what is going on. When upgrading `octokit` to use the `rest` functions, the behavior differs from `octokit/core` in a sense that it will throw 404 errors if no PR is found. With the core version, it will not throw but return an empty array.

The proper fix is to check for the status code, I'll fix you that on the development branch.

Comment (jordan-ae) [reply src/utils/issue.ts:238]: Okay I see. With the proposed `is:pr` fix you sent earlier on the seems to solve the issue. Not sure if it's the most optimal solution but with it the feature works correctly. Including the review delay threshold. You can test it now

**Review by whilefoo — CHANGES_REQUESTED** [review]
I don't think fully implements the specification in which I stated that after the review changes are resolved or the reviewer was requested again, it should be counted as completed

**Comment by jordan-ae** [comment]
> I don't think fully implements the specification in which I stated that after the review changes are resolved or the reviewer was requested again, it should be counted as completed

Missed that when reading the spec. Will add Asap

File: src/handlers/shared/start.ts
Comment (whilefoo) [inline src/handlers/shared/start.ts:166]: if I have 1 issue and 1 PR with requested changes, it will count as 2 here but it should only count 1

File: src/utils/issue.ts
Comment (whilefoo) [inline src/utils/issue.ts:303]: can you also check if all changes are resolved with GraphQL api, so if either a review is requested or changes are solved it should count as approved

File: src/utils/issue.ts
Comment (whilefoo) [inline src/utils/issue.ts:310]: why is it `1m`?

**Review by whilefoo — CHANGES_REQUESTED** [review]

File: src/utils/issue.ts
Comment (jordan-ae) [reply src/utils/issue.ts]: was debugging and mistakenly missed it in the diffs

File: src/utils/issue.ts
Comment (jordan-ae) [reply src/utils/issue.ts]: makes sense I've used paginate

File: src/utils/issue.ts
Comment (jordan-ae) [reply src/utils/issue.ts:293]: Does this mean when we a discussion is resolved it should count as approved?

File: src/utils/issue.ts
Comment (whilefoo) [inline src/utils/issue.ts:229]: you can use `octokit.rest.issues.listEventsForTimeline` instead of hard coded string

File: src/utils/issue.ts
Comment (whilefoo) [inline src/utils/issue.ts:236]: avoid `as`

**Review by whilefoo — CHANGES_REQUESTED** [review]

File: src/utils/issue.ts
Comment (jordan-ae) [reply src/utils/issue.ts]: alright done

File: src/utils/issue.ts
Comment (jordan-ae) [reply src/utils/issue.ts:331]: Done

**Comment by jordan-ae** [comment]
Feature works now but I'm having a hard time writing a handler for this function to make the test pass

**Comment by gentlementlegen** [comment]
The first issue in the tests is
>     HttpError: API rate limit exceeded for 52.238.24.178. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) - https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting

make sure a token is provided during the tests for these.

**Pull request closed without merging**
